File: library.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "library.h"

Library* createLibrary(void) {
    Library* library = (Library*)malloc(sizeof(Library));
    if (library == NULL) return NULL;
    
    library->bookCount = 0;
    for (int i = 0; i < MAX_BOOKS; i++) {
        library->books[i] = NULL;
    }
    
    return library;
}

void destroyLibrary(Library* library) {
    if (library == NULL) return;
    
    for (int i = 0; i < library->bookCount; i++) {
        destroyBook(library->books[i]);
    }
    free(library);
}

int addBook(Library* library, Book* book) {
    if (library == NULL || book == NULL || library->bookCount >= MAX_BOOKS) {
        return 0;
    }
    
    library->books[library->bookCount++] = book;
    return 1;
}

int removeBook(Library* library, const char* isbn) {
    if (library == NULL || isbn == NULL) return 0;
    
    for (int i = 0; i < library->bookCount; i++) {
        if (strcmp(library->books[i]->isbn, isbn) == 0) {
            destroyBook(library->books[i]);
            for (int j = i; j < library->bookCount - 1; j++) {
                library->books[j] = library->books[j + 1];
            }
            library->bookCount--;
            return 1;
        }
    }
    return 0;
}

Book* findBookByISBN(Library* library, const char* isbn) {
    if (library == NULL || isbn == NULL) return NULL;
    
    for (int i = 0; i < library->bookCount; i++) {
        if (strcmp(library->books[i]->isbn, isbn) == 0) {
            return library->books[i];
        }
    }
    return NULL;
}

Book* findBookByTitle(Library* library, const char* title) {
    if (library == NULL || title == NULL) return NULL;
    
    for (int i = 0; i < library->bookCount; i++) {
        if (strstr(library->books[i]->title, title) != NULL) {
            return library->books[i];
        }
    }
    return NULL;
}

void listAllBooks(const Library* library) {
    if (library == NULL) {
        printf("图书馆不存在\n");
        return;
    }
    
    printf("图书馆藏书列表（共%d本）：\n", library->bookCount);
    for (int i = 0; i < library->bookCount; i++) {
        printf("\n--- 第%d本书 ---\n", i + 1);
        printBook(library->books[i]);
    }
}

int borrowBook(Library* library, const char* isbn) {
    Book* book = findBookByISBN(library, isbn);
    if (book == NULL || !isBookAvailable(book)) {
        return 0;
    }
    
    setBookAvailability(book, 0);
    return 1;
}

int returnBook(Library* library, const char* isbn) {
    Book* book = findBookByISBN(library, isbn);
    if (book == NULL || isBookAvailable(book)) {
        return 0;
    }
    
    setBookAvailability(book, 1);
    return 1;
}

int getBookCount(const Library* library) {
    return library != NULL ? library->bookCount : 0;
}

static int compareByTitle(const void* a, const void* b) {
    return strcmp((*(Book**)a)->title, (*(Book**)b)->title);
}

static int compareByAuthor(const void* a, const void* b) {
    return strcmp((*(Book**)a)->author, (*(Book**)b)->author);
}

void sortBooksByTitle(Library* library) {
    if (library == NULL) return;
    qsort(library->books, library->bookCount, sizeof(Book*), compareByTitle);
}

void sortBooksByAuthor(Library* library) {
    if (library == NULL) return;
    qsort(library->books, library->bookCount, sizeof(Book*), compareByAuthor);
} 

File: c2rust/src/library.rs
use crate::book::Book;
use std::cmp::Ordering;

const MAX_BOOKS: usize = 1000;

pub struct Library {
    books: Vec<Book>,
}

impl Library {
    pub fn new() -> Self {
        Library {
            books: Vec::with_capacity(MAX_BOOKS),
        }
    }

    pub fn add_book(&mut self, book: Book) -> bool {
        if self.books.len() >= MAX_BOOKS {
            return false;
        }
        self.books.push(book);
        true
    }

    pub fn remove_book(&mut self, isbn: &str) -> bool {
        if let Some(pos) = self.books.iter().position(|book| book.get_isbn() == isbn) {
            self.books.remove(pos);
            true
        } else {
            false
        }
    }

    pub fn find_book_by_isbn(&self, isbn: &str) -> Option<&Book> {
        self.books.iter().find(|book| book.get_isbn() == isbn)
    }

    pub fn find_book_by_title(&self, title: &str) -> Option<&Book> {
        self.books.iter().find(|book| book.get_title().contains(title))
    }

    pub fn list_all_books(&self) {
        if self.books.is_empty() {
            println!("图书馆暂无藏书");
            return;
        }

        println!("图书馆藏书列表（共{}本）：", self.books.len());
        for (i, book) in self.books.iter().enumerate() {
            println!("\n--- 第{}本书 ---", i + 1);
            book.print();
        }
    }

    pub fn borrow_book(&mut self, isbn: &str) -> bool {
        if let Some(book) = self.books.iter_mut().find(|book| book.get_isbn() == isbn) {
            if book.is_available() {
                book.set_availability(false);
                true
            } else {
                false
            }
        } else {
            false
        }
    }

    pub fn return_book(&mut self, isbn: &str) -> bool {
        if let Some(book) = self.books.iter_mut().find(|book| book.get_isbn() == isbn) {
            if !book.is_available() {
                book.set_availability(true);
                true
            } else {
                false
            }
        } else {
            false
        }
    }

    pub fn get_book_count(&self) -> usize {
        self.books.len()
    }

    pub fn sort_books_by_title(&mut self) {
        self.books.sort_by(|a, b| a.get_title().cmp(b.get_title()));
    }

    pub fn sort_books_by_author(&mut self) {
        self.books.sort_by(|a, b| a.get_author().cmp(b.get_author()));
    }
} 